src/rust/cryptography-x509-validation/src/lib.rs

```diff
@@ -11,6 +11,349 @@ pub mod policy;
+                    }
+                }
+            }
+        }
+
+        if let Some(new_constraints) = new_constraints {
+            self.name_constraints.push(new_constraints);
+        }
+
+        Ok(())
+    }
+}
+
+pub struct Intermediates<'a>(HashSet<Certificate<'a>>);
+
+impl<'a> Intermediates<'a> {
+    fn new(intermediates: impl IntoIterator<Item = Certificate<'a>>) -> Self {
+        Self(intermediates.into_iter().collect())
+    }
+}
+
+pub type Chain<'c> = Vec<Certificate<'c>>;
+
+pub fn verify<'a, 'chain, B: CryptoOps>(
+    leaf: &'a Certificate<'chain>,
+    intermediates: impl IntoIterator<Item = Certificate<'chain>>,
+    policy: &Policy<'_, B>,
+    store: &'a Store<'chain>,
+) -> Result<Chain<'chain>, ValidationError> {
+    let builder = ChainBuilder::new(Intermediates::new(intermediates), policy, store);
+
+    builder.build_chain(leaf)
+}
+
+struct ChainBuilder<'a, 'chain, B: CryptoOps> {
+    intermediates: Intermediates<'chain>,
+    policy: &'a Policy<'a, B>,
+    store: &'a Store<'chain>,
+}
+
+// When applying a name constraint, we need to distinguish between a few different scenarios:
+// * `Applied(true)`: The name constraint is the same type as the SAN and matches.
+// * `Applied(false)`: The name constraint is the same type as the SAN and does not match.
+// * `Skipped`: The name constraint is a different type to the SAN.
+enum ApplyNameConstraintStatus {
+    Applied(bool),
+    Skipped,
+}
+
+impl ApplyNameConstraintStatus {
+    fn is_applied(&self) -> bool {
+        matches!(self, Applied(_))
+    }
+
+    fn is_match(&self) -> bool {
+        matches!(self, Applied(true))
+    }
+}
+
+impl<'a, 'chain, B: CryptoOps> ChainBuilder<'a, 'chain, B> {
+    fn new(
+        intermediates: Intermediates<'chain>,
+        policy: &'a Policy<'a, B>,
+        store: &'a Store<'chain>,
+    ) -> Self {
+        Self {
+            intermediates,
+            policy,
+            store,
+        }
+    }
+
+    fn potential_issuers(
+        &'a self,
+        cert: &'a Certificate<'chain>,
+    ) -> impl Iterator<Item = &'a Certificate<'chain>> + '_ {
+        // TODO: Optimizations:
+        // * Use a backing structure that allows us to search by name
+        //   rather than doing a linear scan
+        // * Search by AKI and other identifiers?
+        self.store
+            .iter()
+            .chain(self.intermediates.0.iter())
+            .filter(|&candidate| candidate.subject() == cert.issuer())
+    }
+
+    fn build_chain_inner(
+        &self,
+        working_cert: &'a Certificate<'chain>,
+        current_depth: u8,
+        working_cert_extensions: &'a Extensions<'chain>,
+        accumulated_constraints: &'a mut AccumulatedNameConstraints<'chain>,
+    ) -> Result<Chain<'chain>, ValidationError> {
+        // Per RFC 5280: Name constraints are not applied
+        // to subjects in self-issued certificates, *unless* the
+        // certificate is the final certificate in the path.
```

Comment:
Not your fault, but language like "final certificate in the path" is totally ambiguous on whether it's the trusted root or EE cert.

Reply:
Thanks for calling that out, this confused me when I changed this around. It refers to the leaf position, so I'll make that clear.